pkg/scheduler/plugins/gang/plugin.go

```diff
@@ -0,0 +1,534 @@
+			return framework.NewStatus(framework.UnschedulableAndUnresolvable, "pod's schedule cycle too large")
+		}
+	}
+	return framework.NewStatus(framework.Success, "")
+}
+
+// PreFilterExtensions returns a PreFilterExtensions interface if the plugin implements one.
+func (p *GangPlugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// PostFilter
+// i. If strict-mode, we will set scheduleCycleValid to false and release all assumed pods.
+// ii. If non-strict mode, we will do nothing.
+func (p *GangPlugin) PostFilter(ctx context.Context, state *framework.CycleState, pod *corev1.Pod,
+	filteredNodeStatusMap framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
+	gangName := getGangNameByPod(pod)
+	if gangName == "" {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Success, "")
+	}
+	gangId := getNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := p.gangCache.getGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName: %v, podName: %v", gangId,
+			getNamespaceSplicingName(pod.Namespace, pod.Name))
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.UnschedulableAndUnresolvable, "can't find gang")
+	}
+
+	if gang.isGangResourceSatisfied() {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Success, "")
+	}
+
+	if gang.getGangMode() == extension.GangModeStrict {
+		p.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			waitingGangId := getNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				getGangNameByPod(waitingPod.GetPod()))
+			if waitingGangId == gangId {
+				klog.Errorf("postFilter rejects the pod, gangName: %v, podName: %v",
+					gangId, getNamespaceSplicingName(waitingPod.GetPod().Namespace, waitingPod.GetPod().Name))
+				waitingPod.Reject(p.Name(), "gang rejection in PostFilter")
+			}
+		})
+		gang.setScheduleCycleValid(false)
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable,
+			fmt.Sprintf("Gang: %v gets rejected this cycle due to Pod: %v is unschedulable even after "+
+				"PostFilter in StrictMode", gangId, pod.Name))
+	}
+
+	return &framework.PostFilterResult{}, framework.NewStatus(framework.Success, "")
+}
+
+// Reserve we add the pod to gang's WaitingForBindChildren map
+func (p *GangPlugin) Reserve(ctx context.Context, state *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
+	gangName := getGangNameByPod(pod)
+	if gangName == "" {
+		return framework.NewStatus(framework.Success, "")
+	}
+	gangId := getNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := p.gangCache.getGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName: %v, podName: %v", gangId,
+			getNamespaceSplicingName(pod.Namespace, pod.Name))
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "can't find gang")
+	}
+	gang.addAssumedPod(pod)
```

Comment:
why not add the waiting Pod in permit phase??

Reply:
In my implement, Permit is used to check the Gang's assuming condition, as for adding the waitingPod to Gang, this action should happen just after the pod being assumed during the Kubernetes ScheduleOne, which is the Reserve stage